File: src/pixel_art_mcp/drawing.py
"""Bounded declarative commands rasterized by the mandatory native pixel helpers."""

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictInt, model_validator

from pixel_art_mcp.pixel_art import Canvas

Symbol = Annotated[str, Field(pattern=r"^[A-Za-z0-9]$")]
PixelRow = Annotated[str, Field(min_length=1, max_length=512, pattern=r"^[A-Za-z0-9.]+$")]
Coordinate = Annotated[StrictInt, Field(ge=0, le=511)]
Dimension = Annotated[StrictInt, Field(ge=1, le=512)]


class PixelModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RepeatedCommand(PixelModel):
    repeat: StrictInt = Field(
        default=1, ge=1, le=128, description="Number of copies, including the first."
    )
    dx: StrictInt = Field(
        default=0, ge=-512, le=512, description="Pixel x offset added for each next copy."
    )
    dy: StrictInt = Field(
        default=0, ge=-512, le=512, description="Pixel y offset added for each next copy."
    )


class Rectangle(RepeatedCommand):
    op: Literal["rect"]
    x: Coordinate
    y: Coordinate
    width: Dimension
    height: Dimension
    color: Symbol

    def bounds(self) -> tuple[int, int, int, int]:
        return self.x, self.y, self.width, self.height

    def cost(self) -> int:
        return self.width * self.height * self.repeat


class Line(RepeatedCommand):
    """One-pixel integer line with inclusive endpoints. Use rectangles for thick posts."""

    op: Literal["line"]
    x1: Coordinate
    y1: Coordinate
    x2: Coordinate
    y2: Coordinate
    color: Symbol

    def bounds(self) -> tuple[int, int, int, int]:
        return (
            min(self.x1, self.x2),
            min(self.y1, self.y2),
            abs(self.x2 - self.x1) + 1,
            abs(self.y2 - self.y1) + 1,
        )

    def cost(self) -> int:
        return max(abs(self.x2 - self.x1), abs(self.y2 - self.y1), 0) * self.repeat + self.repeat


class Stamp(RepeatedCommand):
    """Small literal motif; dots reveal existing pixels, never erase them."""

    op: Literal["stamp"]
    x: Coordinate
    y: Coordinate
    rows: list[PixelRow] = Field(min_length=1, max_length=512)

    @model_validator(mode="after")
    def rectangular(self) -> "Stamp":
        Canvas.from_rows(self.rows)
        return self

    def bounds(self) -> tuple[int, int, int, int]:
        return self.x, self.y, len(self.rows[0]), len(self.rows)

    def cost(self) -> int:
        return len(self.rows[0]) * len(self.rows) * self.repeat


DrawCommand = Annotated[Rectangle | Line | Stamp, Field(discriminator="op")]


class PixelDrawing(PixelModel):
    """Exact integer drawing on a transparent patch, not vector rendering or resampling."""

    width: Dimension
    height: Dimension
    commands: list[DrawCommand] = Field(
        min_length=1,
        max_length=256,
        description="Paint in list order. Each command can repeat with dx/dy; all copies must fit "
        "this patch.",
    )
    mirror_x: StrictBool = Field(
        default=False,
        description="Mirror the completed patch left/right, preserving its dimensions.",
    )

    def cost(self) -> int:
        return sum(command.cost() for command in self.commands)

    def symbols(self) -> set[str]:
        return {
            symbol
            for command in self.commands
            for symbol in ("".join(command.rows) if isinstance(command, Stamp) else command.color)
            if symbol != "."
        }

    @model_validator(mode="after")
    def bounded(self) -> "PixelDrawing":
        if self.cost() > 1_048_576:
            raise ValueError("Drawing exceeds 1048576 paint operations; reduce repetition")
        for index, command in enumerate(self.commands):
            x, y, w, h = command.bounds()
            last_x, last_y = (
                x + command.dx * (command.repeat - 1),
                y + command.dy * (command.repeat - 1),
            )
            if (
                min(x, last_x, y, last_y) < 0
                or max(x, last_x) + w > self.width
                or max(y, last_y) + h > self.height
            ):
                raise ValueError(
                    f"commands[{index}] {command.op} (including repeats) exceeds "
                    f"{self.width}x{self.height} patch; use smaller coordinates/dimensions"
                )
        return self

    def canvas(self) -> Canvas:
        canvas = Canvas(self.width, self.height)
        for command in self.commands:
            for i in range(command.repeat):
                dx, dy = command.dx * i, command.dy * i
                if isinstance(command, Rectangle):
                    canvas.rect(
                        command.x + dx, command.y + dy, command.width, command.height, command.color
                    )
                elif isinstance(command, Line):
                    canvas.line(
                        command.x1 + dx,
                        command.y1 + dy,
                        command.x2 + dx,
                        command.y2 + dy,
                        command.color,
                    )
                else:
                    canvas.stamp(command.x + dx, command.y + dy, command.rows)
        return canvas.mirrored() if self.mirror_x else canvas

File: packages/schema/src/drawing.ts
/**
 * Port of `src/pixel_art_mcp/drawing.py`: bounded declarative commands rasterized by the
 * mandatory native pixel helpers. The actual rasterization (`Canvas.rect`/`.line`/`.stamp`) is
 * the protected core ported in Phase 3 (`packages/pixel-core`, per `docs/typescript-rewrite.md`)
 * -- this module only re-implements the *shape* checks that the authoring contract itself
 * depends on at parse time: `Canvas.from_rows`'s "rows must form a rectangle" check (used by
 * `Stamp.rectangular` and, via `authoring.ts`, `PixelPose.rectangular`), and `PixelDrawing`'s
 * pure-arithmetic cost/bounds accounting (`cost()`, `symbols()`, `bounded()`), none of which
 * require actually painting pixels.
 */

import { z } from "zod";
import {
  addPydanticIssue,
  arrayField,
  intField,
  literalField,
  patternMessage,
  stringField,
  modelObject,
  taggedUnion,
  valueError,
} from "./errors.js";

export const SYMBOL_PATTERN = /^[A-Za-z0-9]$/;
export const PIXEL_ROW_PATTERN = /^[A-Za-z0-9.]+$/;

/** `Symbol = Annotated[str, Field(pattern=r"^[A-Za-z0-9]$")]` */
export const Symbol = stringField({ pattern: SYMBOL_PATTERN, patternText: "^[A-Za-z0-9]$" });

/** `PixelRow = Annotated[str, Field(min_length=1, max_length=512, pattern=r"^[A-Za-z0-9.]+$")]` */
export const PixelRow = stringField({
  minLength: 1,
  maxLength: 512,
  pattern: PIXEL_ROW_PATTERN,
  patternText: "^[A-Za-z0-9.]+$",
});

/** `Coordinate = Annotated[StrictInt, Field(ge=0, le=511)]` */
export const Coordinate = intField({ ge: 0, le: 511 });

/** `Dimension = Annotated[StrictInt, Field(ge=1, le=512)]` */
export const Dimension = intField({ ge: 1, le: 512 });

/**
 * Reproduces `Canvas.from_rows`'s rectangle-shape check (`src/pixel_art_mcp/pixel_art.py`):
 * the actual `Canvas` class is ported in Phase 3, but this one check is invoked directly from
 * the authoring contract's own validators (`Stamp.rectangular`, `PixelPose.rectangular`), so it
 * has to be reproduced here rather than deferred.
 */
export function rectangularRowsError(rows: readonly string[]): string | null {
  if (rows.length === 0 || rows[0]?.length === 0) {
    return "Pixel rows must form a nonempty rectangle; row 0 must not be empty";
  }
  const width = rows[0]?.length ?? 0;
  const bad: [number, number][] = [];
  rows.forEach((row, i) => {
    if (row.length !== width) bad.push([i, row.length]);
  });
  if (bad.length > 0) {
    const details = bad
      .slice(0, 8)
      .map(([i, n]) => `row ${i}: expected ${width}, actual ${n}`)
      .join("; ");
    const remaining = bad.length > 8 ? `; ${bad.length - 8} more mismatched rows` : "";
    return (
      `Pixel rows must form a nonempty rectangle. ${details}${remaining}. ` +
      "Row indices are zero-based; use numeric drawing commands for long shapes."
    );
  }
  return null;
}

/** `RepeatedCommand(PixelModel)` */
const repeatedCommandShape = {
  repeat: intField({ ge: 1, le: 128, description: "Number of copies, including the first." }).default(1),
  dx: intField({ ge: -512, le: 512, description: "Pixel x offset added for each next copy." }).default(0),
  dy: intField({ ge: -512, le: 512, description: "Pixel y offset added for each next copy." }).default(0),
};

export interface RepeatedCommandBounds {
  x: number;
  y: number;
  width: number;
  height: number;
}

export const RectangleSchema = modelObject({
  ...repeatedCommandShape,
  op: literalField(["rect"] as const),
  x: Coordinate,
  y: Coordinate,
  width: Dimension,
  height: Dimension,
  color: Symbol,
});
export type Rectangle = z.infer<typeof RectangleSchema>;

export function rectangleBounds(r: Rectangle): RepeatedCommandBounds {
  return { x: r.x, y: r.y, width: r.width, height: r.height };
}
export function rectangleCost(r: Rectangle): number {
  return r.width * r.height * r.repeat;
}

export const LineSchema = modelObject({
  ...repeatedCommandShape,
  op: literalField(["line"] as const),
  x1: Coordinate,
  y1: Coordinate,
  x2: Coordinate,
  y2: Coordinate,
  color: Symbol,
});
export type Line = z.infer<typeof LineSchema>;

export function lineBounds(l: Line): RepeatedCommandBounds {
  return {
    x: Math.min(l.x1, l.x2),
    y: Math.min(l.y1, l.y2),
    width: Math.abs(l.x2 - l.x1) + 1,
    height: Math.abs(l.y2 - l.y1) + 1,
  };
}
export function lineCost(l: Line): number {
  return Math.max(Math.abs(l.x2 - l.x1), Math.abs(l.y2 - l.y1), 0) * l.repeat + l.repeat;
}

export const StampSchema = modelObject(
  {
    ...repeatedCommandShape,
    op: literalField(["stamp"] as const),
    x: Coordinate,
    y: Coordinate,
    rows: arrayField(PixelRow, { minLength: 1, maxLength: 512 }),
  },
  (value, ctx) => {
    const error = rectangularRowsError(value.rows);
    if (error) valueError(ctx, error);
  },
);
export type Stamp = z.infer<typeof StampSchema>;

export function stampBounds(s: Stamp): RepeatedCommandBounds {
  return { x: s.x, y: s.y, width: s.rows[0]?.length ?? 0, height: s.rows.length };
}
export function stampCost(s: Stamp): number {
  return (s.rows[0]?.length ?? 0) * s.rows.length * s.repeat;
}

export type DrawCommand = Rectangle | Line | Stamp;

export const DrawCommandSchema = taggedUnion("op", {
  rect: RectangleSchema,
  line: LineSchema,
  stamp: StampSchema,
});

export function commandBounds(command: DrawCommand): RepeatedCommandBounds {
  switch (command.op) {
    case "rect":
      return rectangleBounds(command);
    case "line":
      return lineBounds(command);
    case "stamp":
      return stampBounds(command);
  }
}

export function commandCost(command: DrawCommand): number {
  switch (command.op) {
    case "rect":
      return rectangleCost(command);
    case "line":
      return lineCost(command);
    case "stamp":
      return stampCost(command);
  }
}

/** `PixelDrawing(PixelModel)` */
export const PixelDrawingSchema = modelObject(
  {
    width: Dimension,
    height: Dimension,
    commands: arrayField(DrawCommandSchema, {
      minLength: 1,
      maxLength: 256,
      description:
        "Paint in list order. Each command can repeat with dx/dy; all copies must fit " +
        "this patch.",
    }),
    mirror_x: z
      .boolean()
      .default(false)
      .describe("Mirror the completed patch left/right, preserving its dimensions."),
  },
  (value, ctx) => {
    const error = boundedDrawingError(value);
    if (error) valueError(ctx, error);
  },
);
export type PixelDrawing = z.infer<typeof PixelDrawingSchema>;

export function pixelDrawingCost(drawing: PixelDrawing): number {
  return (drawing.commands as DrawCommand[]).reduce((sum, c) => sum + commandCost(c), 0);
}

export function pixelDrawingSymbols(drawing: PixelDrawing): Set<string> {
  const symbols = new Set<string>();
  for (const command of drawing.commands as DrawCommand[]) {
    const chars = command.op === "stamp" ? command.rows.join("").split("") : [command.color];
    for (const symbol of chars) {
      if (symbol !== ".") symbols.add(symbol);
    }
  }
  return symbols;
}

function boundedDrawingError(drawing: PixelDrawing): string | null {
  if (pixelDrawingCost(drawing) > 1_048_576) {
    return "Drawing exceeds 1048576 paint operations; reduce repetition";
  }
  const commands = drawing.commands as DrawCommand[];
  for (let index = 0; index < commands.length; index += 1) {
    const command = commands[index];
    if (!command) continue;
    const { x, y, width: w, height: h } = commandBounds(command);
    const lastX = x + command.dx * (command.repeat - 1);
    const lastY = y + command.dy * (command.repeat - 1);
    if (
      Math.min(x, lastX, y, lastY) < 0 ||
      Math.max(x, lastX) + w > drawing.width ||
      Math.max(y, lastY) + h > drawing.height
    ) {
      return (
        `commands[${index}] ${command.op} (including repeats) exceeds ` +
        `${drawing.width}x${drawing.height} patch; use smaller coordinates/dimensions`
      );
    }
  }
  return null;
}

// Re-exported so callers that only need the literal wording (e.g. tests) don't need to import
// the Zod-issue-shaped helper from errors.ts directly.
export { patternMessage, addPydanticIssue };
